Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp

```diff
@@ -0,0 +1,304 @@
+        return String::formatted("{}", value.as_bool());
+    if (value.is_double())
+        return String::formatted("static_cast<double>({})", value.as_double());
+    if (value.is_array()) {
+        auto const& array = value.as_array();
+        auto child_type = Optional<StringView> {};
+        for (auto const& child_value : array.values()) {
+            if (child_value.is_array())
+                // FIXME: propagate precise cause
+                return Error::from_string_view("Nested arrays are not supported"sv);
+
+#define HANDLE_TYPE(type_name, is_type)                                                             \
+    if (child_value.is_type() && (!child_type.has_value() || child_type.value() == #type_name##sv)) \
+        child_type = #type_name##sv;                                                                \
+    else
+
+            HANDLE_TYPE(StringView, is_string)
+            HANDLE_TYPE(i64, is_integer<i64>)
+            HANDLE_TYPE(u64, is_integer<u64>)
+            HANDLE_TYPE(bool, is_bool)
+            HANDLE_TYPE(double, is_double)
+            // FIXME: propagate precise error cause
+            return Error::from_string_view("Inconsistent contained type in JSON array"sv);
+#undef HANDLE_TYPE
+        }
+        // FIXME: propagate precise error cause
+        if (!child_type.has_value())
+            return Error::from_string_view("Empty JSON array; cannot deduce type."sv);
+
+        StringBuilder initializer;
+        initializer.appendff("Array<{}, {}> {{ "sv, child_type.release_value(), array.size());
+        for (auto const& child_value : array.values())
+            initializer.appendff("{}, ", TRY(generate_initializer_for({}, child_value)));
+        initializer.append("} "sv);
+        return initializer.to_string();
+    }
+    // FIXME: propagate precise cause
+    return Error::from_string_view("Unsupported JSON value"sv);
+}
+
+// Loads an object and assigns it to the RefPtr<Widget> variable named object_name.
+// All loading happens in a separate block.
+static ErrorOr<void> generate_loader_for_object(GUI::GML::Object const& gml_object, SourceGenerator generator, String object_name, size_t indentation, UseObjectConstructor use_object_constructor)
+{
+    generator.set("object_name", object_name.to_deprecated_string());
+    generator.set("class_name", gml_object.name());
+
+    auto append = [&]<size_t N>(auto& generator, char const(&text)[N]) -> ErrorOr<void> {
+        generator.append(TRY(String::repeated(' ', indentation * 4)).bytes_as_string_view());
+        generator.appendln(text);
+        return {};
+    };
+
+    generator.append(TRY(String::repeated(' ', (indentation - 1) * 4)).bytes_as_string_view());
+    generator.appendln("{");
+    if (use_object_constructor == UseObjectConstructor::Yes)
+        TRY(append(generator, "@object_name@ = TRY(@class_name@::try_create());"));
+    else
+        TRY(append(generator, "@object_name@ = TRY(try_make_ref_counted<@class_name@>());"));
```

Comment:
```suggestion
        TRY(append(generator, "@object_name@ = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) @class_name@ ()));"));
```

This would allow us to make the constructors private, preventing weird refcounting errors.

That's not even hypothetical: We used to have these constructors public by default, and a lot of programs constructed refcounted widgets on the stack, causing all kinds of weird things. (Mostly because something took an additional reference, dropped it, the widget saw that its refcount was "again" zero, thought that it should delete itself, called `free()` on stack memory, boom!)

Let's keep those widget constructors private.

Reply:
I am starting to think we should get rid of `make_ref_counted`, since that always requires a public constructor on a refcounted class...

Reply:
That's a good point.

Hmm, maybe a macro instead? The `adopt_nonnull_{ref,own}_or_enomem` incantation appears in surprisingly many places. Either way, let's avoid public constructors here, and a different PR could solve that problem. If you're not feeling like it, tell me, then I'll give it a stab :D

Reply:
I'll use your suggestion for now though, you're right.